pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -11,50 +11,116 @@ import (
 // - perform any pod upgrade (left for rolling upgrade phase)
 func HandleUpscaleAndSpecChanges(
-	k8sClient k8s.Client,
-	es v1alpha1.Elasticsearch,
-	scheme *runtime.Scheme,
-	expectedResources nodespec.ResourcesList,
+	ctx upscaleCtx,
 	actualStatefulSets sset.StatefulSetList,
+	expectedResources nodespec.ResourcesList,
 ) error {
-	// TODO: there is a split brain possibility here if going from 1 to 3 masters or 3 to 7.
-	//  We should add only one master node at a time for safety.
-	//  See https://github.com/elastic/cloud-on-k8s/issues/1281.
-
-	for _, nodeSpecRes := range expectedResources {
-		// always reconcile config (will apply to new & recreated pods)
-		if err := settings.ReconcileConfig(k8sClient, es, nodeSpecRes.StatefulSet.Name, nodeSpecRes.Config); err != nil {
+	// adjust expected replicas to control nodes creation and deletion
+	adjusted, err := adjustReplicas(ctx, actualStatefulSets, expectedResources)
+	if err != nil {
+		return err
+	}
+	// reconcile all resources
+	for _, res := range adjusted {
+		if err := settings.ReconcileConfig(ctx.k8sClient, ctx.es, res.StatefulSet.Name, res.Config); err != nil {
 			return err
 		}
-		if _, err := common.ReconcileService(k8sClient, scheme, &nodeSpecRes.HeadlessService, &es); err != nil {
+		if _, err := common.ReconcileService(ctx.k8sClient, ctx.scheme, &res.HeadlessService, &ctx.es); err != nil {
 			return err
 		}
-		ssetToApply := *nodeSpecRes.StatefulSet.DeepCopy()
-		actual, alreadyExists := actualStatefulSets.GetByName(ssetToApply.Name)
-		if alreadyExists {
-			ssetToApply = adaptForExistingStatefulSet(actual, ssetToApply)
+		if err := sset.ReconcileStatefulSet(ctx.k8sClient, ctx.scheme, ctx.es, res.StatefulSet); err != nil {
+			return err
 		}
-		if err := sset.ReconcileStatefulSet(k8sClient, scheme, es, ssetToApply); err != nil {
+	}
+	return nil
+}
+
+func adjustReplicas(
+	ctx upscaleCtx,
+	actualStatefulSets sset.StatefulSetList,
+	expectedResources nodespec.ResourcesList,
+) (nodespec.ResourcesList, error) {
+	adjustedResources := make(nodespec.ResourcesList, 0, len(expectedResources))
+	for _, nodeSpecRes := range expectedResources {
+		adjustedSset, err := adjustStatefulSetReplicas(ctx, actualStatefulSets, nodeSpecRes.StatefulSet)
+		if err != nil {
+			return nil, err
+		}
+		nodeSpecRes.StatefulSet = adjustedSset
+		adjustedResources = append(adjustedResources, nodeSpecRes)
+	}
+	// adapt resources configuration to match adjusted replicas
+	if err := adjustZenConfig(ctx, adjustedResources); err != nil {
+		return nil, err
+	}
+	return adjustedResources, nil
+}
+
+func adjustZenConfig(ctx upscaleCtx, resources nodespec.ResourcesList) error {
+	// patch configs to consider zen1 minimum master nodes
+	if err := zen1.SetupMinimumMasterNodesConfig(resources); err != nil {
+		return err
+	}
+	// patch configs to consider zen2 initial master nodes if cluster is not bootstrapped yet
+	if !AnnotatedForBootstrap(ctx.es) {
```

Comment:
I think we might have an issue here on upgrades of single node clusters from 6.x to 7.x. Even though the cluster has technically been bootstrapped by rolling the only master node we create a situation where the node expects to join an existing cluster when coming back because initial master nodes are not set see #1642

Reply:
Wow, good catch!

Reply:
In other terms, we may need to differentiate:
* a cluster bootstrapped in the sense that this is not the initial cluster creation
* a cluster bootstrapped with Zen2, in the sense it has seen some initial master nodes already

Reply:
I think we already had this problem before?

Reply:
Also related: https://github.com/elastic/cloud-on-k8s/issues/1281#issuecomment-524832319

Reply:
Also related: https://github.com/elastic/cloud-on-k8s/issues/1281#issuecomment-524832319

Reply:
> I think we already had this problem before?

Yes absolutely it is nothing introduce in this PR, just an FYI

Reply:
Created https://github.com/elastic/cloud-on-k8s/issues/1660